File: katas/java/Patterns/strategy/StrategyTests.java
package Patterns.strategy;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests para el patrón Strategy
 * Valida que el patrón Strategy permita cambiar algoritmos en tiempo de ejecución
 */
public class StrategyTests {
    
    @Test
    public void testPaymentProcessor_WithCreditCardStrategy() {
        // Arrange
        PaymentStrategy creditCard = new CreditCardPayment("1234-5678-9012-3456", "John Doe");
        PaymentProcessor processor = new PaymentProcessor(creditCard);
        
        // Act
        String result = processor.processPayment(100.0);
        
        // Assert
        assertNotNull(result, "Debe procesar el pago con tarjeta de crédito");
        assertTrue(result.contains("Credit Card") || result.contains("tarjeta"), 
                   "El resultado debe mencionar tarjeta de crédito");
        assertTrue(result.contains("100"), "Debe incluir el monto");
    }
    
    @Test
    public void testPaymentProcessor_WithPayPalStrategy() {
        // Arrange
        PaymentStrategy paypal = new PayPalPayment("user@email.com");
        PaymentProcessor processor = new PaymentProcessor(paypal);
        
        // Act
        String result = processor.processPayment(250.0);
        
        // Assert
        assertNotNull(result, "Debe procesar el pago con PayPal");
        assertTrue(result.contains("PayPal"), "El resultado debe mencionar PayPal");
        assertTrue(result.contains("250"), "Debe incluir el monto");
    }
    
    @Test
    public void testPaymentProcessor_WithBitcoinStrategy() {
        // Arrange
        PaymentStrategy bitcoin = new BitcoinPayment("1A2B3C4D5E6F7G8H9I0J");
        PaymentProcessor processor = new PaymentProcessor(bitcoin);
        
        // Act
        String result = processor.processPayment(500.0);
        
        // Assert
        assertNotNull(result, "Debe procesar el pago con Bitcoin");
        assertTrue(result.contains("Bitcoin"), "El resultado debe mencionar Bitcoin");
        assertTrue(result.contains("500"), "Debe incluir el monto");
    }
    
    @Test
    public void testPaymentProcessor_CanChangeStrategy() {
        // Arrange
        PaymentProcessor processor = new PaymentProcessor(new CreditCardPayment("1234", "John"));
        
        // Act - Cambiar de estrategia en tiempo de ejecución
        String result1 = processor.processPayment(100.0);
        processor.setPaymentStrategy(new PayPalPayment("user@email.com"));
        String result2 = processor.processPayment(200.0);
        
        // Assert
        assertTrue(result1.contains("Credit Card") || result1.contains("tarjeta"), 
                   "Primera transacción debe usar tarjeta de crédito");
        assertTrue(result2.contains("PayPal"), "Segunda transacción debe usar PayPal");
    }
    
    @Test
    public void testCreditCardPayment_ContainsCardInfo() {
        // Arrange
        CreditCardPayment payment = new CreditCardPayment("1234-5678-9012-3456", "Jane Doe");
        
        // Act
        String result = payment.pay(150.0);
        
        // Assert
        assertTrue(result.contains("1234") || result.contains("****"), 
                   "Debe incluir información de la tarjeta (enmascarada o visible)");
        assertTrue(result.contains("Jane Doe") || result.contains("Jane"), 
                   "Debe incluir el nombre del titular");
    }
    
    @Test
    public void testPayPalPayment_ContainsEmail() {
        // Arrange
        PayPalPayment payment = new PayPalPayment("test@example.com");
        
        // Act
        String result = payment.pay(75.0);
        
        // Assert
        assertTrue(result.contains("test@example.com"), "Debe incluir el email de PayPal");
    }
    
    @Test
    public void testBitcoinPayment_ContainsWalletAddress() {
        // Arrange
        BitcoinPayment payment = new BitcoinPayment("1BvBMSEYstWetqTFn5Au4m4GFg7xJaNVN2");
        
        // Act
        String result = payment.pay(1000.0);
        
        // Assert
        assertTrue(result.contains("1BvBM") || result.contains("wallet"), 
                   "Debe incluir información de la dirección de wallet");
    }
}

File: katas/php/Patterns/strategy/StrategyTests.php
<?php

use PHPUnit\Framework\TestCase;

require_once 'strategy-good.php';

/**
 * Tests para el patrón Strategy
 * Valida que el patrón Strategy permita cambiar algoritmos en tiempo de ejecución
 */
class StrategyTests extends TestCase
{
    public function testPaymentProcessor_WithCreditCardStrategy()
    {
        // Arrange
        $creditCard = new CreditCardPayment("1234-5678-9012-3456", "John Doe");
        $processor = new PaymentProcessor($creditCard);
        
        // Act
        $result = $processor->processPayment(100.0);
        
        // Assert
        $this->assertNotNull($result, 'Debe procesar el pago con tarjeta de crédito');
        $this->assertTrue(
            str_contains($result, 'Credit Card') || str_contains($result, 'tarjeta'),
            'El resultado debe mencionar tarjeta de crédito'
        );
        $this->assertStringContainsString("100", $result, 'Debe incluir el monto');
    }
    
    public function testPaymentProcessor_WithPayPalStrategy()
    {
        // Arrange
        $paypal = new PayPalPayment("user@email.com");
        $processor = new PaymentProcessor($paypal);
        
        // Act
        $result = $processor->processPayment(250.0);
        
        // Assert
        $this->assertNotNull($result, 'Debe procesar el pago con PayPal');
        $this->assertStringContainsString("PayPal", $result, 'El resultado debe mencionar PayPal');
        $this->assertStringContainsString("250", $result, 'Debe incluir el monto');
    }
    
    public function testPaymentProcessor_WithBitcoinStrategy()
    {
        // Arrange
        $bitcoin = new BitcoinPayment("1A2B3C4D5E6F7G8H9I0J");
        $processor = new PaymentProcessor($bitcoin);
        
        // Act
        $result = $processor->processPayment(500.0);
        
        // Assert
        $this->assertNotNull($result, 'Debe procesar el pago con Bitcoin');
        $this->assertStringContainsString("Bitcoin", $result, 'El resultado debe mencionar Bitcoin');
        $this->assertStringContainsString("500", $result, 'Debe incluir el monto');
    }
    
    public function testPaymentProcessor_CanChangeStrategy()
    {
        // Arrange
        $processor = new PaymentProcessor(new CreditCardPayment("1234", "John"));
        
        // Act - Cambiar de estrategia en tiempo de ejecución
        $result1 = $processor->processPayment(100.0);
        $processor->setPaymentStrategy(new PayPalPayment("user@email.com"));
        $result2 = $processor->processPayment(200.0);
        
        // Assert
        $this->assertTrue(
            str_contains($result1, 'Credit Card') || str_contains($result1, 'tarjeta'),
            'Primera transacción debe usar tarjeta de crédito'
        );
        $this->assertStringContainsString("PayPal", $result2, 'Segunda transacción debe usar PayPal');
    }
    
    public function testCreditCardPayment_ContainsCardInfo()
    {
        // Arrange
        $payment = new CreditCardPayment("1234-5678-9012-3456", "Jane Doe");
        
        // Act
        $result = $payment->pay(150.0);
        
        // Assert
        $this->assertTrue(
            str_contains($result, '1234') || str_contains($result, '****'),
            'Debe incluir información de la tarjeta (enmascarada o visible)'
        );
        $this->assertTrue(
            str_contains($result, 'Jane Doe') || str_contains($result, 'Jane'),
            'Debe incluir el nombre del titular'
        );
    }
    
    public function testPayPalPayment_ContainsEmail()
    {
        // Arrange
        $payment = new PayPalPayment("test@example.com");
        
        // Act
        $result = $payment->pay(75.0);
        
        // Assert
        $this->assertStringContainsString("test@example.com", $result, 'Debe incluir el email de PayPal');
    }
    
    public function testBitcoinPayment_ContainsWalletAddress()
    {
        // Arrange
        $payment = new BitcoinPayment("1BvBMSEYstWetqTFn5Au4m4GFg7xJaNVN2");
        
        // Act
        $result = $payment->pay(1000.0);
        
        // Assert
        $this->assertTrue(
            str_contains($result, '1BvBM') || str_contains($result, 'wallet'),
            'Debe incluir información de la dirección de wallet'
        );
    }
}
